### ISSUE TITLE
metadata-proxy not able to handle too many pods/node

### ISSUE THREAD

**Issue by shyamjvs** [kubernetes/kubernetes#55695]
Kubemark-500 started failing continuously from ~10 days ago (https://k8s-gubernator.appspot.com/builds/kubernetes-jenkins/logs/ci-kubernetes-kubemark-500-gce?before=9575).
On digging up a bit, I found out that quite frequently the hollow-nodes are failing to create/delete pods. And there are many such logs from hollow-kubelet:

```
I1114 08:08:24.861541       7 metadata.go:212] Failed to Get service accounts from gce metadata server: Get http://metadata.google.internal./computeMetadata/v1/instance/service-accounts/: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
I1114 08:08:25.114072       7 metadata.go:212] Failed to Get service accounts from gce metadata server: Get http://metadata.google.internal./computeMetadata/v1/instance/service-accounts/: net/http: request canceled (Client.Timeout exceeded while awaiting headers)
I1114 08:12:51.032702       7 metadata.go:212] Failed to Get service accounts from gce metadata server: Get http://metadata.google.internal./computeMetadata/v1/instance/service-accounts/: dial tcp 169.254.169.254:80: getsockopt: connection refused
```

(taken from https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-kubemark-500-gce/9697/artifacts/kubemark-500-minion-group-5f7n/kubelet-hollow-node-68krn.log)

This seems clearly related to changes to metadata-proxy. More specifically change https://github.com/kubernetes/kubernetes/pull/54653 in the [commit diff](https://github.com/kubernetes/kubernetes/compare/7d759ba0c...d4d326b39).

I checked from the currently running kubemark-500 cluster and found that the metadata-proxy pods are crashing frequently:

```
$ kubectl get pods -n kube-system -l k8s-app=metadata-proxy
NAME                        READY     STATUS             RESTARTS   AGE
metadata-proxy-v0.1-6tnk7   2/2       Running            11         1h
metadata-proxy-v0.1-fmkbd   1/2       CrashLoopBackOff   11         1h
metadata-proxy-v0.1-gp8zp   2/2       Running            11         1h
metadata-proxy-v0.1-hvqlt   1/2       CrashLoopBackOff   11         1h
metadata-proxy-v0.1-p45p5   2/2       Running            7          1h
metadata-proxy-v0.1-pkn6m   2/2       Running            7          1h
metadata-proxy-v0.1-scmxs   2/2       Running            0          1h
metadata-proxy-v0.1-sdccv   2/2       Running            11         1h
```

And the reason for them crashing seems to be OOM-killing:

```
$ kubectl describe pods -n kube-system metadata-proxy-v0.1-gp8zp
Name:           metadata-proxy-v0.1-gp8zp
Namespace:      kube-system
Node:           kubemark-500-minion-group-g6h7/10.142.0.6
Start Time:     Tue, 14 Nov 2017 09:26:12 +0000
...
Containers:
  metadata-proxy:
    Container ID:   docker://c52477c5fa28cb12d9b3b2dd5dd196654f8a2e291a64ce54910e0fa479372634
    Image:          gcr.io/google_containers/metadata-proxy:v0.1.4
    Image ID:       docker-pullable://gcr.io/google_containers/metadata-proxy@sha256:9ca02cc773e286b28942103e2482ce74e281ef2c18936610d379b26ecb89c533
    Port:           <none>
    State:          Waiting
      Reason:       CrashLoopBackOff
    Last State:     Terminated
      Reason:       OOMKilled
      Exit Code:    137
```

cc @kubernetes/sig-node-bugs @kubernetes/sig-scalability-misc @porridge

**Comment by shyamjvs** [comment]
I'm noticing this issue only on kubemark-500 and NOT on kubemark-100 and kubemark-5000. And my hypothesis is that it's because the no. of hollow-nodes running per real node is high in kubemark-500 compared to kubemark-100 and kubemark-5000:

for kubemark-100, #nodes=3 => ~33 hollow-nodes per node
for kubemark-500, #nodes=7 => ~72 hollow-nodes per node
for kubemark-5000, #nodes=80 => ~63 hollow-nodes per node

This means that a single metadata-proxy pod is not able to handle the load coming from 72 pods trying to access metadata.
Basically **metadata-proxy isn't scaling with #pods (that're trying to access metadata) per node**.

I'll try to confirm this hypothesis by tweaking the #real-nodes and checking.

**Comment by shyamjvs** [comment]
Seems like my hypothesis is right. I started a kubemark 1000-node cluster locally on 15 nodes (i.e ~67 hollow-nodes/node) and am seeing those errors on all nodes. But on 16 nodes (i.e ~63 hollow-nodes/node) they're only seen on one node, on which there are more than avg no. of hollow-nodes (66). With 17 nodes, I'm not seeing it at all.

Few options we have here:
- revert the offending PR #54653 and fix metadata-proxy before bringing back the change: IMO this is the best option
- increase memory request/limit of metadata-proxy: This can affect the node allocatable capacity and might be undesirable for small nodes
- increase the #nodes for kubemark-500 job: I don't think we should be doing this as it is not solving the real problem and we'd still be violating "k8s supports 100 pods/node"

**Comment by shyamjvs** [comment]
cc @kubernetes/sig-node-bugs @ihmccreery @dchen1107

**Comment by porridge** [comment]
I think we should roll back and only bring the change back after
demonstrating that the issue is fixed.
Also, if memory usage depends on the number of requests in flight, then
metadata proxy should be load-tested to the standard that metadata server
itself is (which I believe is on the order of a few thousand qps).

**Comment by shyamjvs** [comment]
Yes, I agree. We need load-testing of metadata proxy.
I sent out the revert PR for restoring back to healthy state, so we don't need to live with the regression until a 'well-tested' fix rolls in.

**Comment by shyamjvs** [comment]
I took a quick look at #54653. Seems like there is this change that earlier metadata-proxy had 32MB of memory request & limit totally for itself. After introducing prometheus-to-sd sidecar, that was split evenly across both (i.e 16MB each). This is likely the reason that metadata-proxy is being memory starved now.

Re-adjusting the split could help here. Thoughts?

**Comment by shyamjvs** [comment]
Seems like PR #55715 fixed this. I filed a separate issue #55797 for load-testing metadata-proxy.
/close
